syntax = "proto3";
import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

option go_package = "./syslog";
package syslog;

enum Severity {
  Emergency = 0;      ///< System is unusable.
  Alert = 1;          ///< Action must be taken immediately.
  Critical = 2;       ///< Critical condition.
  Error = 3;          ///< Error condition.
  Warning = 4;        ///< Warning condition.
  Notice = 5;         ///< Normal but significant condition.
  Informational = 6;  ///< Informational message.
  Debug = 7;           ///< Debug message.
}

message SyslogDataDefinition {
  int64 identity = 1; ///< Unique ID of the data definition
  string name = 2; ///< Internal name of the value. Ident/name
  string display_name = 3; ///< Friendly name of the value.
  string description = 4; ///< Descriptive text
  string unit = 5; ///< Unit of measure
}

message SyslogDataValue {
  int64 identity = 1; ///< Unique database ID of the structured data definition.
  string name = 2;    ///< Value name (display name)
  string value = 3;   ///< Data value
  string unit = 4;    ///< Unit
}

message SyslogMessage {
  int64 identity = 1;                       ///< Unique ID (database index) of the event
  Severity severity = 2;                    ///< Severity code.
  uint32 facility = 3;                      ///< Facility code. Normally not used
  google.protobuf.Timestamp timestamp = 4;  ///< Nano-second since 1970.
  string text = 5;                          ///< Event message text
  string hostname = 6;                      ///< Host name. Display name or name.
  string application_name = 7;              ///< Application name
  string process_id = 8;                    ///< Process ID (PID). Not really used.
  string message_id = 9;                    ///< Message ID
  repeated SyslogDataValue data_values = 10; ///< Extra data/value pair.
}

message EventMessage {
  int64 identity = 1;    ///< Unique ID (database index) of the event
  Severity severity = 2; ///< Severity code.
  google.protobuf.Timestamp timestamp = 3;  ///< Nano-second since 1970.
  string text = 4;       ///< Event message text
}

message SyslogFilter {
  optional Severity level = 1; ///< Severity level. Lower severity is filtered out.
  optional uint32 facility = 2; ///< Facility code.
  string text_filter = 3; ///< Wild card filter of the text.
  optional google.protobuf.Timestamp from_time = 4; ///< From time.
  optional google.protobuf.Timestamp to_time = 5; ///< From time.
  repeated SyslogDataDefinition data = 6; ///< Shows these data definition values
  optional int64 from_id = 7; ///< If set. Filter out all lower id.
  uint64 offset = 8; ///< From row offset.
  optional uint32 count = 9; ///< Number of rows returned.
}

message SyslogCount {
  uint64 count = 1;
}

service SyslogService {
  rpc GetLastEvent(google.protobuf.Empty) returns (EventMessage) {}
  rpc GetCount(SyslogFilter) returns (SyslogCount) {}
  rpc GetEvent(SyslogFilter) returns (stream EventMessage) {}
  rpc GetSyslog(SyslogFilter) returns (stream SyslogMessage) {}
  rpc GetDataDefinitions(google.protobuf.Empty) returns (stream SyslogDataDefinition) {}
  rpc AddNewMessage(SyslogMessage) returns (google.protobuf.Empty) {}

}